Modules/offline_awareness/_offline_aware_abc.py

```diff
@@ -0,0 +1,118 @@
+"""
+_offline_aware_abc.py - Provides an offline awareness ABC
+
+Copyright (c) 2018 The Fuel Rats Mischief,
+All rights reserved.
+
+Licensed under the BSD 3-Clause License.
+
+See LICENSE.md
+"""
+
+import logging
+import weakref
+from abc import abstractmethod, ABC
+from typing import List, NoReturn
+
+LOG = logging.getLogger(f"mecha.{__name__}")
+
+
+class OfflineAwareABC(ABC):
+    """
+    Interface for offline-aware classes
+    """
+    __slots__ = []
+    _storage: List[weakref.finalize] = []
+    """
+    Private weak reference storage
+    """
+
+    online = False
+    """
+    Are we in online mode?
+    """
+
+    def __init__(self):
+        """
+        registers the subclass obj instance weakly. used for event propagation
+        """
+
+        # create a finalizer-based weak reference, tie the callback to our GC method
+        # then append it to our storage object
+        self._storage.append(weakref.finalize(self, self.__gc))
+
+    @abstractmethod
+    async def on_online(self) -> NoReturn:
+        """
+        on_online event callback, invoked  when the system moves to "online" mode
+        """
+
+    @abstractmethod
+    async def on_offline(self) -> NoReturn:
+        """
+        on_offline event callback, invoked when the system moves to "offline" mode
+        """
+
+    @classmethod
+    async def go_online(cls) -> NoReturn:
```

Comment:
Please log when this function is invoked.

Reply:
see above

Reply:
resolved in 

0ab407a